### PR TITLE
Fix error handling in fn:doc-available

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/1364

### CODE REVIEW

File: src/org/exist/xquery/functions/fn/FunDocAvailable.java [hunk 1]
```diff
@@ -35,79 +33,83 @@
...
+        if (!arg.isEmpty()) {
             final String path = arg.itemAt(0).getStringValue();
             try {
                 result = BooleanValue.valueOf(DocUtils.isDocumentAvailable(this.context, path));
             } catch (final XPathException e) {
-                e.prependMessage(ErrorCodes.FODC0005, "");
-                throw e;
+                if (e.getCause() instanceof SAXParseException) {
+                    result = BooleanValue.FALSE;
+                } else if(e.getMessage().contains("is a binary resource, not an XML document")) {
```
Comment (shabanovd) [inline src/org/exist/xquery/functions/fn/FunDocAvailable.java]: very bad hack

Comment (adamretter) [inline src/org/exist/xquery/functions/fn/FunDocAvailable.java]: I completely agree. However there is no better fix without either changing the API (which we can't do for a hotfix release) or duplicating code. I think a better fix could come in future. This was just designed for 3.1.1 and it improves the current situation.

Comment (dizzzz) [inline src/org/exist/xquery/functions/fn/FunDocAvailable.java]: :-)  can't we check in the EXpathException for an ErrorCode ?

Comment (shabanovd) [inline src/org/exist/xquery/functions/fn/FunDocAvailable.java]: yes, my proposal to use ErrorCode too

File: test/src/xquery/fn.xql [hunk 2]
```diff
@@ -21,4 +41,17 @@ declare
 function fnt:normalize-unicode($normalization-form as xs:string) {
     count(string-to-codepoints(normalize-unicode($fnt:composed, $normalization-form))),
     count(string-to-codepoints(normalize-unicode($fnt:decomposed, $normalization-form)))
-};
\ No newline at end of file
+};
+
+declare
```
Comment (shabanovd) [inline test/src/xquery/fn.xql]: no test for external invalid xml resource?

File: src/org/exist/xquery/functions/fn/FunDocAvailable.java (hunk 1 above)
Comment (adamretter) [inline src/org/exist/xquery/functions/fn/FunDocAvailable.java]: There is no difference between checking a constant error code and and a constant string here, both have the same fragility.

The solution is to refactor the code in `DocUtils` to be more granular and differentiate between different kinds of errors, however that cannot be done without API changes. This fix is just a temporary fix that improves the current state.

File: test/src/xquery/fn.xql (hunk 2 above)
Comment (adamretter) [inline test/src/xquery/fn.xql]: There was no test before. I added the tests that I needed. Feel free to send a PR with more tests if you want them.

File: src/org/exist/xquery/functions/fn/FunDocAvailable.java (hunk 1 above)
Comment (shabanovd) [inline src/org/exist/xquery/functions/fn/FunDocAvailable.java]: Huge difference, because _constant_ error have clear where-is-used. No needs for API change!

File: test/src/xquery/fn.xql (hunk 2 above)
Comment (shabanovd) [inline test/src/xquery/fn.xql]: @adamretter you want to close #1364 but do not cover it by test

Comment (adamretter) [inline test/src/xquery/fn.xql]: How do you propose I test that without introducing an external dependency?

Comment (shabanovd) [inline test/src/xquery/fn.xql]: There is already number of it ... Just use url from issue report.

Comment (adamretter) [inline test/src/xquery/fn.xql]: @shabanovd That would introduce a dependency on an external website into the test suite. The problem is that if that website is offline our test suite fails. So I don't want an external dependency.

Comment (shabanovd) [inline test/src/xquery/fn.xql]: another option to add corrupted xml and reference it by `http://localhost:...`

Comment (shabanovd) [inline test/src/xquery/fn.xql]: BTW, [here](https://github.com/eXist-db/exist/blob/f7643dcb5d3c68c8c695b73d1ede6b0f8a8820b8/test/src/org/exist/xquery/XQueryTest.java#L1219) tests that have an external website dependency

Comment (joewiz) [inline test/src/xquery/fn.xql]: @shabanovd Good idea to reference some local file that is corrupted.

A more roundabout, but attractive, approach would be to submit a test to qt3, and then have eXist run this test along with all the others. (We'd need to get the qt3 runner working again though...)

Comment (adamretter) [inline test/src/xquery/fn.xql]: @shabanovd The XQuery Test Suite runner operates on an embedded XML Database, so there is no Jetty webserver.

Comment (adamretter) [inline test/src/xquery/fn.xql]: @shabanovd No tests should have external web dependencies as it has caused us pain in the past. That above `org.exist.xquery.XQueryTest` should be fixed in a separate PR.

Comment (shabanovd) [inline test/src/xquery/fn.xql]: remove `Closes 1364` and then it good to go

Comment (adamretter) [inline test/src/xquery/fn.xql]: Why? The PR does fix the issue in 1364.

Comment (shabanovd) [inline test/src/xquery/fn.xql]: where do you have test for available _external_ resource but invalid xml?

Comment (adamretter) [inline test/src/xquery/fn.xql]: I don't have a test for that which you know from our above discussion. However this does fix the issue, which you could confirm by testing yourself.
